docs/usage/model_config.md

```diff
@@ -164,3 +167,28 @@ For example:
 {!.tmp_examples/model_config_alias_precedence.py!}
 ```
 _(This script is complete, it should run "as is")_
+
+## Smart Union
+
+By default, if you set a `Union` type for a field, _pydantic_ tries to validate (and coerce if it can) in the order of the union,
+which can be unexpected in some cases
```

Comment:
Is it worth adding the warning here from #2839, because python treats unions as singletons so the first definition anywhere in the import chain is what is used and makes that behaviour even more unexpected
so something like

```suggestion

By default, if you set a `Union` type for a field, _pydantic_ tries to validate (and coerce if it can) in the order of the union.

!!! warning
    `typing.Unions` also ignore order when they are [defined](https://docs.python.org/3/library/typing.html#typing.Union),
    so `Union[int, float] == Union[float, int]` which can lead to unexpected behaviour when combined with matching based on the `Union` type order.
    Please note that this can also be affected by third party libraries and their internal type definitions and the import orders.
```